evm-fungible-token/src/lib.rs

```diff
@@ -0,0 +1,351 @@
+                "proof": proof_1,
+            })
+            .to_string()
+            .as_bytes(),
+            NO_DEPOSIT,
+            prepaid_gas / 4,
+        );
+        env::promise_return(promise1);
+    }
+
+    /// Finish depositing once the proof was successfully validated.
+    /// Can only be called by the contract itself.
+    #[private]
+    pub fn finish_deposit(&mut self, new_owner_id: AccountId, amount: U128, proof: Proof) {
+        log!(
+            "finish_deposit - Promise results: {:?}",
+            env::promise_results_count()
+        );
+        log!("Amount: {:?}", amount);
+        assert_eq!(env::promise_results_count(), 1);
+        let data0: Vec<u8> = match env::promise_result(0) {
+            PromiseResult::Successful(x) => x,
+            _ => panic!("Promise with index 0 failed"),
+        };
+        let verification_success: bool = serde_json::from_slice(&data0).unwrap();
+        assert!(verification_success, "Failed to verify the proof");
+        self.record_proof(&proof.get_key());
+
+        self.mint(new_owner_id, amount.into());
+    }
+
+    /// Mint Fungible Token for account
+    /// TODO: should be related to NEP-145
+    #[private]
+    fn mint(&mut self, owner_id: AccountId, amount: Balance) {
+        log!("Mint {:?} tokens for: {:?}", amount, owner_id);
+        if self.token.accounts.get(&owner_id).is_none() {
+            self.token.accounts.insert(&owner_id, &amount);
+        } else {
+            self.token.internal_deposit(&owner_id, amount);
+        }
+    }
+
+    /// Burn Fungible Token for account
+    #[private]
+    fn burn(&mut self, owner_id: AccountId, amount: Balance) {
+        log!("Burn {:?} tokens for: {:?}", amount, owner_id);
+        self.token.internal_withdraw(&owner_id, amount);
+    }
+
+    #[payable]
+    pub fn withdraw(&mut self, proof: Proof) {
+        log!("Start withdraw");
+        // let event = EthWithdrawEvent::from_log_entry_data(&proof.log_entry_data);
+        //================================
+        // TODO: for testing only
+        let event = EthWithdrawEvent {
+            eth_custodian_address: self.eth_custodian_address,
+            sender: "sender1".into(),
+            amount: U128::from(5),
+            recipient: "rcv1".into(),
+            fee: U128::from(2),
+        };
+        //================================
+
+        assert_eq!(
+            event.eth_custodian_address,
+            self.eth_custodian_address,
+            "Event's address {} does not match custodian address of this token {}",
+            hex::encode(&event.eth_custodian_address),
+            hex::encode(&self.eth_custodian_address),
+        );
+        let proof_1 = proof.clone();
+        let account_id = env::current_account_id();
+        let prepaid_gas = env::prepaid_gas();
+        log!("Withdraw verify_log_entry");
+        let promise0 = env::promise_create(
+            self.prover_account.clone(),
+            b"verify_log_entry",
```

Comment:
I think verify proof in deposit and withdraw should use different method. If i understand correctly,
deposit: have a proof of EthDepositedEvent from near-evm contract, use it to mint in evm-fungible-token
withdraw: have a proof of EthWithdrawEvent from evm-fungible-token, use it to withdraw token locked in near-evm contract. Also it seems this PR doesn't implement the withdraw part for with near-evm token, I think it should either returns a proof (so user can submit the proof to withdraw) or cross contract call near-evm to withdraw

Reply:
That's true. Withdraw should not use the prover at all. Discussed this with @mrLSD yesterday.